Add Check Support for Java 17 Sealed Classes: Indentation

child of #14969 :

documentation: Add Check Support for Java 17 Sealed Classe


---
```
PS D:\CS\test> javac src/Test.java                                              
PS D:\CS\test> cat config.xml                                                   
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
               <module name="Indentation">

        </module>
    </module>
</module>
PS D:\CS\test> cat src/Test.java                                                
     sealed   // violation
class Test  // violation
    permits A, B { // violation
}

    final class A extends Test { // violation
}

   sealed  // violation
   class B extends Test // violation
permits // violation
       D {
}

    non-sealed // violation
class D // violation
    extends B { // violation
}

Starting audit...
[ERROR] D:\CS\test\src\Test.java:1:3: 'class def modifier' has incorrect indentation level 2, expected level should be 0. [Indentation]
[ERROR] D:\CS\test\src\Test.java:2:1: 'class' has incorrect indentation level 0, expected level should be 6. [Indentation]
[ERROR] D:\CS\test\src\Test.java:3:5: 'permits' has incorrect indentation level 4, expected level should be 6. [Indentation]
[ERROR] D:\CS\test\src\Test.java:6:5: 'class def modifier' has incorrect indentation level 4, expected level should be 0. [Indentation]
[ERROR] D:\CS\test\src\Test.java:9:4: 'class def modifier' has incorrect indentation level 3, expected level should be 0. [Indentation]
[ERROR] D:\CS\test\src\Test.java:10:4: 'class' has incorrect indentation level 3, expected level should be 7. [Indentation]
[ERROR] D:\CS\test\src\Test.java:11:1: 'permits' has incorrect indentation level 0, expected level should be 7. [Indentation]
[ERROR] D:\CS\test\src\Test.java:15:5: 'class def modifier' has incorrect indentation level 4, expected level should be 0. [Indentation]
[ERROR] D:\CS\test\src\Test.java:16:1: 'class' has incorrect indentation level 0, expected level should be 8. [Indentation]
[ERROR] D:\CS\test\src\Test.java:17:5: 'extends' has incorrect indentation level 4, expected level should be 8. [Indentation]
Audit done.
Checkstyle ends with 10 errors.

PS D:\CS\test> 
```

---

Describe what you want in detail.

we are good here. The check works correctly on all new tokens

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/indentation/IndentationCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/indentation/IndentationCheckTest.java
@@ -3091,6 +3091,39 @@ public class IndentationCheckTest extends AbstractModuleTestSupport {
         verifyWarns(checkConfig, fileName, expected);
     }
 
+    @Test
+    public void testIndentationSealedClasses()
+            throws Exception {
+        final DefaultConfiguration checkConfig = createModuleConfig(IndentationCheck.class);
+        checkConfig.addProperty("forceStrictCondition", "true");
+        checkConfig.addProperty("tabWidth", "4");
+        checkConfig.addProperty("basicOffset", "4");
+        checkConfig.addProperty("braceAdjustment", "0");
+        checkConfig.addProperty("caseIndent", "4");
+        checkConfig.addProperty("throwsIndent", "8");
+
+        final String fileName = getNonCompilablePath(
+                "InputIndentationSealedClasses.java");
+        final String[] expected = {
+            "14:1: " + getCheckMessage(MSG_ERROR, "class def modifier", 0, 4),
+            "15:2: " + getCheckMessage(MSG_ERROR, "class", 1, 4),
+            "16:6: " + getCheckMessage(MSG_ERROR, "permits", 5, 4),
+            "19:5: " + getCheckMessage(MSG_ERROR, "class", 4, 8),
+            "20:5: " + getCheckMessage(MSG_ERROR, "permits", 4, 8),
+            "28:1: " + getCheckMessage(MSG_ERROR, "class def modifier", 0, 4),
+            "29:9: " + getCheckMessage(MSG_ERROR, "extends", 8, 4),
+            "32:5: " + getCheckMessage(MSG_ERROR, "extends", 4, 8),
+            "38:5: " + getCheckMessage(MSG_ERROR, "class", 4, 8),
+            "39:1: " + getCheckMessage(MSG_ERROR, "permits", 0, 8),
+            "40:13: " + getCheckMessage(MSG_ERROR, "C", 12, 8),
+            "48:5: " + getCheckMessage(MSG_ERROR, "class", 4, 8),
+            "49:5: " + getCheckMessage(MSG_ERROR, "C", 4, 8),
+            "55:1: " + getCheckMessage(MSG_ERROR, "class def modifier", 0, 4),
+            "56:9: " + getCheckMessage(MSG_ERROR, "class", 8, 4),
+        };
+        verifyWarns(checkConfig, fileName, expected);
+    }
+
     private static final class IndentAudit implements AuditListener {
 
         private final IndentComment[] comments;

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationSealedClasses.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationSealedClasses.java
@@ -1,0 +1,72 @@
+//non-compiled with javac: Compilable with Java21                           //indent:0 exp:0
+package com.puppycrawl.tools.checkstyle.checks.indentation.indentation;     //indent:0 exp:0
+
+
+/* Config:                                                                  //indent:0 exp:0
+ * basicOffset = 4                                                          //indent:1 exp:1
+ * braceAdjustment = 0                                                      //indent:1 exp:1
+ * caseIndent = 4                                                           //indent:1 exp:1
+ * tabWidth = 4                                                             //indent:1 exp:1
+ * throwsIndent = 8                                                         //indent:1 exp:1
+ * forceStrictCondition = true                                              //indent:1 exp:1
+ */                                                                         //indent:1 exp:1
+public class InputIndentationSealedClasses { //indent:0 exp:0
+sealed //indent:0 exp:4 warn
+ class Test //indent:1 exp:4 warn
+     permits A, B { } //indent:5 exp:4 warn
+
+    sealed //indent:4 exp:4
+    class Test2 //indent:4 exp:8 warn
+    permits X, Y, FF { } //indent:4 exp:8 warn
+
+    sealed //indent:4 exp:4
+        class Test3 //indent:8 exp:8
+        permits W, Z { } //indent:8 exp:8
+
+    sealed class A extends Test permits D, DD, E, F, EE, O { } //indent:4 exp:4
+
+final class D //indent:0 exp:4 warn
+        extends A { } //indent:8 exp:4 warn
+
+    final class DD //indent:4 exp:4
+    extends A { } //indent:4 exp:8 warn
+
+    final class O //indent:4 exp:4
+        extends A { } //indent:8 exp:8
+
+    sealed //indent:4 exp:4
+    class B extends Test //indent:4 exp:8 warn
+permits //indent:0 exp:8 warn
+            C, CC { } //indent:12 exp:8 warn
+
+    sealed //indent:4 exp:4
+        class FF extends Test2 //indent:8 exp:8
+        permits //indent:8 exp:8
+        T { } //indent:8 exp:8
+
+    non-sealed //indent:4 exp:4
+    class //indent:4 exp:8 warn
+    C extends B { } //indent:4 exp:8 warn
+
+    non-sealed //indent:4 exp:4
+        class //indent:8 exp:8
+        CC extends B { } //indent:8 exp:8
+
+non-sealed //indent:0 exp:4 warn
+        class E extends A { } //indent:8 exp:4 warn
+
+    non-sealed //indent:4 exp:4
+        class EE extends A { } //indent:8 exp:8
+
+    sealed class F extends A //indent:4 exp:4
+        permits R { } //indent:8 exp:8
+
+    final class R extends F { } //indent:4 exp:4
+
+    final class X extends Test2 { } //indent:4 exp:4
+    final class Y extends Test2 { } //indent:4 exp:4
+    final class W extends Test3 { } //indent:4 exp:4
+    final class Z extends Test3 { } //indent:4 exp:4
+    final class T extends FF { } //indent:4 exp:4
+
+} //indent:0 exp:0

